WebGPUTable API

ilijapuaca opened the pull request:
Revisited `AttributeDescriptor` usage. Initial `WebGPUTable` API formalization

ibgreen commented on cpp/modules/deck.gl/core/src/lib/attribute/attribute-descriptor.h:
```diff
@@ -18,37 +18,32 @@
+  using AttributeBuilder = auto(const std::shared_ptr<arrow::Table>&) -> std::shared_ptr<arrow::Array>;
+  AttributeDescriptor(const std::string& name, const wgpu::VertexFormat format,
+                      std::function<AttributeBuilder> attributeBuilder)
+      : lumagl::AttributeDescriptor{name, format}, attributeBuilder{attributeBuilder} {}
```
What if we just used arrow types here. We would not need to expose wgpu at all?

(You may need another field to record whether this is an instanced attribute. It can just be a boolean, but maps to https://gpuweb.github.io/gpuweb/#enumdef-gpuinputstepmode.)

ibgreen commented on cpp/modules/deck.gl/core/src/lib/attribute/attribute-manager.cc:
```diff
@@ -49,20 +49,21 @@ void AttributeManager::invalidateAll() {
   // Create an empty output table
-  std::vector<std::shared_ptr<arrow::Field>> fields{};
-  std::vector<std::shared_ptr<WebGPUColumn>> columns{};
+  std::vector<std::shared_ptr<WebGPUField>> fields{};
```
- Does each table need a specific type of fields and schemas? 
- Or can we reuse the arrow types?

ibgreen commented on cpp/modules/deck.gl/core/src/lib/attribute/attribute-manager.h:
```diff
@@ -61,7 +61,7 @@ class AttributeManager {
   void setNeedsRedraw();
 
-  void add(const std::shared_ptr<lumagl::AttributeDescriptor>& descriptor);
+  void add(const std::shared_ptr<AttributeDescriptor>& descriptor);
```
I  am not sure it should be possible to add and remove attributes (this would require rebuilding the model pipeline) so this could be an array in the constructor.

ibgreen replied:
Attribute descriptors should be passed by value and copied. No need to manage them with shared pointers!

ibgreen commented on cpp/modules/deck.gl/layers/src/line-layer/line-layer.cpp:
```diff
@@ -163,7 +160,7 @@ auto LineLayer::getSourcePositionData(const std::shared_ptr<arrow::Table>& table
   }
 
-  return ArrowMapper::mapVector3DoubleColumn(table, props->getSourcePosition);
+  return ArrowMapper::mapVector3FloatColumn(table, props->getSourcePosition);
```
This is fine for now, but eventually we are going to need to split doubles into two floats...

D1 D2 D3 => F1high F1low F2high F2 low...

ibgreen commented on cpp/modules/deck.gl/core/src/arrow/arrow-utils.cc:
```diff
@@ -0,0 +1,90 @@
+      return arrow::fixed_size_list(arrow::float32(), 2);
+    case Format::Float3:
+      return arrow::fixed_size_list(arrow::float32(), 3);
+    case Format::Float4:
```
I was disappointed to see that there was no support for double precision.

ibgreen commented on cpp/modules/luma.gl/webgpu/src/data/webgpu-schema.h:
```diff
@@ -0,0 +1,61 @@
+ public:
+  explicit WebGPUSchema(const std::vector<std::shared_ptr<WebGPUField>>& fields) : _fields{fields} {}
+
+  /// \brief Returns number of fields that this schema contains.
```
My initial idea was to make this "API compatible" with Arrow https://arrow.apache.org/docs/cpp/api/datatype.html#classarrow_1_1_schema.

Use the same methods, and method naming... 

Right now, this is inspired by schema but offers a slightly different set of methods. Would we loose much by aligning the methods?

It could also mean using PascalCase instead of camelCase in this class - I could go either way on that.

ilijapuaca commented:
Addressed the comments and refined the API further per discussion from today. @ibgreen if possible take a look at the last commit and I can address any potential comments
